src/assets/asset-manager.ts

```diff
@@ -0,0 +1,323 @@
+  }
+
+  static async addVirtualAsset(
+    handle: File | FileSystemFileHandle
+  ): Promise<void> {
+    let file: File
+    if ('getFile' in handle) {
+      file = await handle.getFile()
+    } else {
+      file = handle
+    }
+
+    await fs.promises.writeFile(
+      file.name,
+      new Uint8Array(await file.arrayBuffer()),
+      {flush: true}
+    )
+
+    await this._updateMap()
+    this._emitChangeEvent()
+  }
+
+  static async renameVirtualAsset(
+    oldName: string,
+    newName: string
+  ): Promise<void> {
+    if (this.filesystemType !== 'virtual') {
+      return
+    }
+
+    await fs.promises.rename(oldName, newName)
+
+    await this._updateMap()
+    this._emitChangeEvent()
+  }
+
+  static async deleteVirtualAsset(name: string): Promise<void> {
+    if (this.filesystemType !== 'virtual') {
+      return
+    }
+
+    await fs.promises.unlink(name)
+
+    await this._updateMap()
+    this._emitChangeEvent()
+  }
+
+  static async clearVirtualAssets(): Promise<void> {
+    if (this.filesystemType !== 'virtual' || !this._assetMap) {
+      return
+    }
+
+    const removals = []
+    for (const name of Object.keys(this._assetMap)) {
+      removals.push(fs.promises.unlink(name))
+    }
+
+    await Promise.all(removals)
+
+    await this._updateMap()
+    this._emitChangeEvent()
+  }
+
+  private static async _mountRoot(fs: FileSystem): Promise<void> {
+    this._unmountRoot()
+    ZenFS.mount('/', fs)
+    await this._updateMap()
+  }
+
+  private static _unmountRoot(): void {
+    if (ZenFS.mounts.has('/')) {
+      ZenFS.umount('/')
+      this._archiveHandle = undefined
+      this._directoryHandle = undefined
+    }
+  }
+
+  private static async _updateMap(): Promise<void> {
+    this._clearAssetMap()
+    if (!ZenFS.mounts.has('/')) {
+      console.debug(LOG_TAG, 'Root filesystem unmounted, skipping _updateMap')
+      return
+    }
+    const entries: string[] = []
+    entries.push(...(await fs.promises.readdir('/')))
+    while (entries.length > 0) {
+      const entry = entries.shift()!
+      // remove the leading / present in some backends
+      const name = entry.replace(/^\//, '')
+      const stat = await fs.promises.stat(entry)
+      if (stat.isDirectory()) {
+        entries.unshift(
+          ...(await fs.promises.readdir(entry)).map(child => {
+            if (child.startsWith('/')) return child
+            return `${entry}/${child}`
+          })
+        )
+      } else {
+        const file = await fs.promises.open(entry)
+        const buffer = await file.readFile()
+        await file.close()
+        const basename = entry.split('/').at(-1)!
+        const extension = basename.split('.').at(-1)!
+        const mimetype = MIME[extension] ?? DEFAULT_MIME
+        this._assetMap![name] = window.URL.createObjectURL(
+          new Blob([buffer], {type: mimetype})
+        )
+        this._assetCount++
+      }
+    }
+    console.debug(LOG_TAG, `Found ${this._assetCount} files`, this._assetMap)
```

Comment:
I'll remove the debug logs and change the two mount logs to warnings since they may be relevant to users:
```'Failed to remount the previously used directory'```
``` 'Failed to remount the previously used ZIP file' ```